### PR TITLE
Product review schema with descriptions. 

### PR DESCRIPTION
**Pull request by nrkapoor**
GraphQL schema for product reviews and ratings GraphQL functionality.

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/architecture?pullRequest=298) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/architecture?pullRequest=298) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: paliarush<br/>:x: nrkapoor

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+type ProductReviewRatingMetadata {
+    id: String! @doc(description: "Base 64 encoded id of the rating category name. for e.g. price, quality etc.")
+    name: String! @doc(description: "The review rating name for e.g. quality, price etc.")
+    values: [ProductReviewRatingValueMetadata!]! @doc(description: "Sorted based on position")
+}
+
+type ProductReviewRatingValueMetadata {
+    value_id: String! @doc(description: "Base 64 encoded id of the value id.")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    value_id: String! @doc(description: "Base 64 encoded rating value id.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    productReviewRatingsMetadata(): ProductReviewRatingsMetadata! @doc(description: "Metadata required by clients to render ratings & reviews section.")
+}
+
+type ProductReviewRatingsMetadata {
+    items: [ProductReviewRatingMetadata!]! @doc(description: "Sorted based on position")
+}
+
+type ProductReviewRatingMetadata {
+    id: String! @doc(description: "Base 64 encoded id of the rating category name. for e.g. price, quality etc.")
+    name: String! @doc(description: "The review rating name for e.g. quality, price etc.")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    name: String! @doc(description: "The review rating name for example quality, price")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+type Query {
+    productReviewRatingsMetadata(): ProductReviewRatingsMetadata! @doc(description: "Metadata required by clients to render ratings & reviews section.")
+}
+
+type ProductReviewRatingsMetadata {
+    items: [ProductReviewRatingMetadata!]! @doc(description: "Sorted based on position")
+}
+
+type ProductReviewRatingMetadata {
+    id: String! @doc(description: "Base 64 encoded id of the rating category name. for e.g. price, quality etc.")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    id: String! @doc(description: "Base 64 encoded rating id.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    value: Int! @doc(description: "The rating value given by customer. Possible values: 1 to 5")
+}
+
+
+type Query {
+    productReviewRatingsMetadata(): ProductReviewRatingsMetadata! @doc(description: "Metadata required by clients to render ratings & reviews section.")
+}
+
+type ProductReviewRatingsMetadata {
+    items: [ProductReviewRatingMetadata!]! @doc(description: "Sorted based on position")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    items: [ProductReviewRatingMetadata!]! @doc(description: "List of product reviews sorted based on position")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+type ProductReviewRatingMetadata {
+    id: String! @doc(description: "Base 64 encoded id of the rating category name. for e.g. price, quality etc.")
+    name: String! @doc(description: "The review rating name for e.g. quality, price etc.")
+    values: [ProductReviewRatingValueMetadata!]! @doc(description: "Sorted based on position")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    values: [ProductReviewRatingValueMetadata!]! @doc(description: "List of product review ratings sorted based on position.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    summary: String! @doc(description: "The review summary (a.k.a title")
+    text: String! @doc(description: "The review text.")
+    nickname: String! @doc(description: "The customer's nickname. Defaults to customer name if logged in.")
+    created_at: String! @doc(description: "Date indicating when the review was created.")
+    average_rating: Float! @doc(description: "The average rating for product review.")
+    ratings_breakdown: [ProductReviewRating!]! @doc(description: "An array of ratings by rating category. for e.g. quality, price etc.")
+}
+
+type ProductReviewRating {
+    name: String! @doc(description: "The review rating name for e.g. quality, price etc.")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    name: String! @doc(description: "The review rating name for example quality, price.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+type ProductReview @doc(description: "Details of a product review") {
+    product: ProductInterface! @doc(description: "Contains details about the reviewed product")
+    summary: String! @doc(description: "The review summary (a.k.a title")
+    text: String! @doc(description: "The review text.")
+    nickname: String! @doc(description: "The customer's nickname. Defaults to customer name if logged in.")
+    created_at: String! @doc(description: "Date indicating when the review was created.")
+    average_rating: Float! @doc(description: "The average rating for product review.")
+    ratings_breakdown: [ProductReviewRating!]! @doc(description: "An array of ratings by rating category. for e.g. quality, price etc.")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]: Please check https://www.grammarly.com/blog/know-your-latin-i-e-vs-e-g/

```suggestion
    ratings_breakdown: [ProductReviewRating!]! @doc(description: "An array of ratings by rating category. For example quality, price.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    review_count: Int! @doc(description: "The total count of all the reviews given to the product.")
+    reviews(
+        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once. This attribute is optional."),
+        currentPage: Int = 1! @doc(description: "Specifies which page of results to return. The default value is 1."),
+    ): ProductReviews!
+}
+
+type ProductReviews {
+    items: [ProductReview]! @doc(description: "And array of product reviews.")
+    page_info: SearchResultPageInfo! @doc(description: "An object that includes the page_info and currentPage values specified in the query.")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    page_info: SearchResultPageInfo! @doc(description: "Metadata for pagination rendering.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    average_review_rating: Float! @doc(description: "The average of all the ratings given to the product.")
+    review_count: Int! @doc(description: "The total count of all the reviews given to the product.")
+    reviews(
+        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once. This attribute is optional."),
+        currentPage: Int = 1! @doc(description: "Specifies which page of results to return. The default value is 1."),
+    ): ProductReviews!
+}
+
+type ProductReviews {
+    items: [ProductReview]! @doc(description: "And array of product reviews.")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    items: [ProductReview]! @doc(description: "An array of product reviews.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        currentPage: Int = 1! @doc(description: "Specifies which page of results to return. The default value is 1."),
+    ): ProductReviews!
+}
+
+type ProductInterface {
+    average_review_rating: Float! @doc(description: "The average of all the ratings given to the product.")
+    review_count: Int! @doc(description: "The total count of all the reviews given to the product.")
+    reviews(
+        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once. This attribute is optional."),
+        currentPage: Int = 1! @doc(description: "Specifies which page of results to return. The default value is 1."),
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
        currentPage: Int = 1 @doc(description: "Specifies which page of results to return."),
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    ): ProductReviews!
+}
+
+type ProductInterface {
+    average_review_rating: Float! @doc(description: "The average of all the ratings given to the product.")
+    review_count: Int! @doc(description: "The total count of all the reviews given to the product.")
+    reviews(
+        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once. This attribute is optional."),
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once."),
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 12]
```diff
@@ -0,0 +1,76 @@
+type Customer {
+    reviews(
+        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once. This attribute is optional."),
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
        pageSize: Int = 20 @doc(description: "Specifies the maximum number of results to return at once."),
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        currentPage: Int = 1! @doc(description: "Specifies which page of results to return. The default value is 1."),
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
        currentPage: Int = 1 @doc(description: "Specifies which page of results to return."),
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+input CreateProductReviewInput {
+    sku: String! @doc(description: "The SKU of the product that the review is assigned")
+    nickname: String! @doc(description: "The customer's nickname. Defaults to customer name if logged in.")
+    summary: String! @doc(description: "The review summary (a.k.a title")
+    text: String! @doc(description: "The review text.")
+    ratings: [ProductReviewRatingInput!]! @doc(description: "Ratings details by category. e.g price: 5, quality: 4 etc")
+}
+
+type ProductReviewRatingInput {
+    id: String! @doc(description: "Base 64 encoded id of the value id.")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    id: String! @doc(description: "Base 64 encoded rating id.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    value_id: String! @doc(description: "e.g Good, Perfect, 3, 4, 5")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    value_id: String! @doc(description: "Base 64 encoded rating value id.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+    text: String! @doc(description: "The review text.")
+    nickname: String! @doc(description: "The customer's nickname. Defaults to customer name if logged in.")
+    created_at: String! @doc(description: "Date indicating when the review was created.")
+    average_rating: Float! @doc(description: "The average rating for product review.")
+    ratings_breakdown: [ProductReviewRating!]! @doc(description: "An array of ratings by rating category. for e.g. quality, price etc.")
+}
+
+type ProductReviewRating {
+    name: String! @doc(description: "The review rating name for e.g. quality, price etc.")
+    value: Int! @doc(description: "The rating value given by customer. Possible values: 1 to 5")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]:
```suggestion
    value: Int! @doc(description: "The rating value given by customer. Possible values by default: 1 to 5.")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    text: String! @doc(description: "The review text.")
+    nickname: String! @doc(description: "The customer's nickname. Defaults to customer name if logged in.")
+    created_at: String! @doc(description: "Date indicating when the review was created.")
+    average_rating: Float! @doc(description: "The average rating for product review.")
+    ratings_breakdown: [ProductReviewRating!]! @doc(description: "An array of ratings by rating category. For example quality, price.")
+}
+
+type ProductReviewRating {
+    name: String! @doc(description: "The review rating name for example quality, price.")
+    value: Int! @doc(description: "The rating value given by customer. Possible values by default: 1 to 5.")
```
Comment (buskamuza) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]: What is "value", in relation to the metadata? Probably should be "string".

Comment (paliarush) [reply design-documents/graph-ql/coverage/product-reviews.graphqls]: Yes, I think you are right.
@nrkapoor please change it and then we can merge the PR.

**Pull request merged**